Add support for Renesas RA CEU (Capture Engine Unit) for RA8M1 and RA8D1

thaoluonguw opened the pull request:
Add support for Renesas RA CEU (Capture Engine Unit):

- Add video driver use CEU
- Add dts for CEU on RA8M1 and RA8D1
- Enable CEU support for EK-RA8D1
- Add support for EK-RA8D1 on Video capture to LVGL sample.
It works with OV7670 camera, and MIPI graphics expansion board included in the Kit

thaoluonguw committed:
boards: renesas: ek_ra8d1: Update document for using Camera

Add SW1 configuration for using Camera Exapansion Port (J59)

Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>

josuah reviewed (changes requested):
Thank you for implementing the Renesas video driver!

The driver appears in a good shape WRT video APIs, with a few key points to modify (marked with :x), and a few proposal to improve it further, non-blocking (i.e. could be improved over time eventually, although appreciated if modified now).

Let me know if anything needs clarifications.

khanhnguyenwz committed:
drivers: video: add support for Renesas RA CEU driver

Add support for the Renesas RA Capture Engine Unit (CEU),
including driver source files, Kconfig options, and DTS bindings.

- Add initial implementation of the RA CEU driver
- Add dedicated Kconfig and CMake integration
- Provide Devicetree bindings for the RA CEU
- Update module Kconfig to include the new driver

This enables image capture functionality using the CEU peripheral
on Renesas RA series MCUs.

Signed-off-by: Duy Vo <duy.vo.xc@bp.renesas.com>
Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
include: dt-bindings: renesas: Add RA CEU pin definitions

Add pin definitions required by the RA Capture Engine Unit

Signed-off-by: Duy Vo <duy.vo.xc@bp.renesas.com>
Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
dts: arm: renesas: Add CEU nodes for RA8D1 and RA8M1 SoCs

Add CEU to r7fa8d1xh.dtsi and r7fa8m1xh.dtsi

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
boards: renesas: Enable CEU video capture support on EK-RA8D1

- Add CEU pin configuration to ek_ra8d1-pinctrl.dtsi
- Enable CEU node and Arducam 20-pin connector in ek_ra8d1.dts
- Configure PWM3 as external XCLK via pwm-clock node
- Update board YAML to declare video support

Signed-off-by: Duy Vo <duy.vo.xc@bp.renesas.com>
Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
boards: shields: Support OV7670 DVP 20-pin shield on EK-RA8D1 board

Support OV7670 DVP 20-pin shield on the EK-RA8D1 board

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
samples: video: Add CEU support for Renesas RA boards

Add EK-RA8D1 board support for capture and capture_to_lvgl samples

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

josuah commented on samples/drivers/video/capture/sample.yaml:
```diff
@@ -11,6 +11,7 @@ tests:
       - platform:mimxrt1170_evk/mimxrt1176/cm7:SHIELD="nxp_btb44_ov5640;rk055hdmipi4ma0"
       - platform:mimxrt1170_evk@B/mimxrt1176/cm7:SHIELD="nxp_btb44_ov5640;rk055hdmipi4ma0"
       - platform:frdm_mcxn947/mcxn947/cpu0:SHIELD="dvp_20pin_ov7670;lcd_par_s035_8080"
+      - platform:ek_ra8d1:SHIELD="dvp_20pin_ov7670;rtkmipilcdb00000be"
```
[blocking :x:]
It is possible that the current CI failure happens as the `sample.yaml` platform definition need the board listed with all the qualifiers. Maybe something like this but am not familiar enough with this chip to be entirely sure:

```suggestion
      - platform:ek_ra8d1/r7fa8d1bhecbd:SHIELD="dvp_20pin_ov7670;rtkmipilcdb00000be"
```

josuah reviewed (commented):
Thank you for the modifications!
I have "folded" the points it does resolve to allow you to keep track of feedback.
A few items left might need some adjustment (i.e. for CI), but I am looking forward to see this merged!

josuah commented on drivers/video/video_renesas_ra_ceu.c:
```diff
@@ -0,0 +1,460 @@
+		.fsp_ctrl = &video_renesas_ra_ceu_fsp_ctrl##inst,                                  \
+		.fsp_cfg = &video_renesas_ra_ceu_fsp_cfg##inst,                                    \
+		.fsp_extend_cfg = &video_renesas_ra_ceu_fsp_extend_cfg##inst,                      \
+	};                                                                                         \
```
I am marking this as resolved as it is only a method to reduce RAM usage by a few bytes and is not expected to have impact.

josuah commented on drivers/video/video_renesas_ra_ceu.c:
```diff
@@ -0,0 +1,460 @@
+                                                                                                   \
+	PINCTRL_DT_INST_DEFINE(inst);                                                              \
+                                                                                                   \
+	static struct video_renesas_ra_ceu_config video_renesas_ra_ceu_config##inst = {            \
```
Solved by the latest commit, thank you!

josuah commented on drivers/video/video_renesas_ra_ceu.c:
```diff
@@ -0,0 +1,460 @@
+	fsp_err_t err;
+	int ret;
+
+	fmt->pitch = fmt->width * video_bits_per_pixel(fmt->pixelformat) / BITS_PER_BYTE;
```
This now moved to the bottom of the function so am resolving this thread.
Thanks all.

josuah commented:
No problem, it is possible to pause activity on PRs for as long as you need.

The last item I think is missing would be adding it to the [`build_all`](http://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all/video) Twister test to make it easy to bulk test that all drivers still build well despite API changes.

Here is an example of how it can be done for `renesas_ra_ceu`: 

```diff
diff --git a/tests/drivers/build_all/video/testcase.yaml b/tests/drivers/build_all/video/testcase.yaml
index c7a5e07161e..1851a1d99e7 100644
--- a/tests/drivers/build_all/video/testcase.yaml
+++ b/tests/drivers/build_all/video/testcase.yaml
@@ -36,3 +36,8 @@ tests:
       - stm32n6570_dk/stm32n657xx/sb
     extra_args:
       - platform:stm32n6570_dk/stm32n657xx/sb:SHIELD=st_b_cams_imx_mb1854
+  drivers.video.renesas_ra_ceu.build:
+    platform_allow:
+      - ek_ra8d1/r7fa8d1bhecbd
+    extra_args:
+      - platform:ek_ra8d1/r7fa8d1bhecbd:SHIELD=dvp_20pin_ov7670
```

<details>

<summary>
<code>west twister -s drivers.video.renesas_ra_ceu.build</code>
</summary>

<pre><code>Renaming output directory to /home/tinyvision/zephyrproject/zephyr/twister-out.3
INFO    - Using Ninja..
INFO    - Zephyr version: v4.2.0-rc1-110-g356db7abcffd
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per testsuite scenario
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 4
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   1  100%  built (not run):    1, filtered:    0, failed:    0, error:    0
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 0 of 1 executed test configurations passed (0.00%), 1 built (not run), 0 failed, 0 errored, with no warnings in 56.95 seconds.
INFO    - 0 of 0 executed test cases passed (0.00%) on 0 out of total 1115 platforms (0.00%).
INFO    - 1 selected test cases not executed: 1 not run (built only).
INFO    - 0 test configurations executed on platforms, 1 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
</code></pre>

</details>

ngphibang commented on drivers/video/video_renesas_ra_ceu.c:
```diff
@@ -0,0 +1,460 @@
+	}
+}
+
+static int video_renesas_ra_ceu_get_format(const struct device *dev, struct video_format *fmt)
```
@thaoluonguw : This is a small point but do you think it's better to shorten the function name a bit ? The prefix `video_renesas_` seems not needed as all these functions are static, no worry that other functions will have the same name

josuah replied:
I think there is a `_csi` variant coming at some point. This would become `ra_csi_get_format` then I guess...

<img width="1172" height="179" alt="scrot_20250711_142621_1172x179" src="https://github.com/user-attachments/assets/20a57772-49e4-4381-9bcd-d3122474c9fe" />

thenguyenyf replied:
As my viewpoint, better to keep the `video_renesas_ra_*` prefix for consistency with the name-spacing rule. With the `*_ceu` surfix, as mentioned by @josuah , an `*_csi` variant driver will be sent out later, so let's keep it to specific hardware function.

ngphibang commented on drivers/video/video_renesas_ra_ceu.c:
```diff
@@ -0,0 +1,538 @@
+#include <zephyr/drivers/video.h>
+#include <zephyr/irq.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <errno.h>
```
It seems better to sort the headers byte alphabetical order ?

```
/* Common header */
#include <errno.h>
#include <soc.h>

/* Zephyr header */

#include "r_ceu.h"
#include "video_device.h"
```

ngphibang commented on dts/bindings/video/renesas,ra-ceu.yaml:
```diff
@@ -0,0 +1,83 @@
+    description: |
+      Specifies the data transfer unit size to the bus bridge module.
+
+child-binding:
+  child-binding:
+    include: video-interfaces.yaml
+
+    properties:
+      hsync-active:
+        required: true
+
+      vsync-active:
+        required: true
+
+      pclk-sample:
+        required: true
+
+      vsync-sample:
+        required: true
+        type: int
+        enum:
+          - 0 # Falling edge
+          - 1 # Rising edge
+        description: |
+          Sample on the falling or rising edge of the vsync signal.
+
+      hsync-sample:
+        required: true
+        type: int
+        enum:
+          - 0 # Falling edge
+          - 1 # Rising edge
+        description: |
+          Sample on the falling or rising edge of the hsync signal.
+
+      bus-width:
```
Yes, this is not entirely a redefinition, except for some properties. I think it's OK for `bus-width`, `vsync-sample` and `hsync-sample`.

About `hsync-active`, `vsync-active` and `pclk-sample`, a redefinition to says these properties are mandatory, that is OK. But I don't know, in Zephyr, by doing that, is the previous definitions are overridden by the last one so we lost the `enum` attributes ?

ngphibang reviewed (changes requested):
Could you add a test case in build_all so that the driver build can be triggered by CI ?

```
diff --git a/tests/drivers/build_all/video/testcase.yaml b/tests/drivers/build_all/video/testcase.yaml
index c7a5e07161e..f8c03137bfa 100644
--- a/tests/drivers/build_all/video/testcase.yaml
+++ b/tests/drivers/build_all/video/testcase.yaml
@@ -36,3 +36,8 @@ tests:
       - stm32n6570_dk/stm32n657xx/sb
     extra_args:
       - platform:stm32n6570_dk/stm32n657xx/sb:SHIELD=st_b_cams_imx_mb1854
+  drivers.video.renesas_ra_ceu.build:
+    platform_allow:
+      - ek_ra8d1/r7fa8d1bhecbd
+    extra_args:
+      - platform:ek_ra8d1/r7fa8d1bhecbd:SHIELD="dvp_20pin_ov7670;rtkmipilcdb00000be"
```

khoa-nguyen-18 commented on boards/renesas/ek_ra8d1/ek_ra8d1.dts:
```diff
@@ -255,6 +266,21 @@
 	};
 };
 
+&pwm3 {
+	pinctrl-0 = <&pwm3_default>;
+	pinctrl-names = "default";
+	interrupts = <51 12>, <52 12>;
+	interrupt-names = "gtioca", "overflow";
+
+	cam_clock: pwmclock {
+		compatible = "pwm-clock";
+		status = "disabled";
+		#clock-cells = <1>;
+		clock-frequency = <24000000>;
+		pwms = <&pwm3 0 PWM_KHZ(24000) PWM_POLARITY_NORMAL>;
+	};
+};
+
```
Should we define this `pwm3` and `cam_clock` in the shield layer? I feel it's wasting two interrupt lines on the board when they're not used, and the cam_clock properties are specific to each shield.

khoa-nguyen-18 commented on boards/renesas/ek_ra8d1/ek_ra8d1-pinctrl.dtsi:
```diff
@@ -297,4 +310,22 @@
 		};
 	};
+
+	ceu_default: ceu_default {
+		group1 {
+			/* CEU */
+			psels = <RA_PSEL(RA_PSEL_CEU, 4, 0)>, /* VIO_D0 */
+				<RA_PSEL(RA_PSEL_CEU, 4, 1)>, /* VIO_D1 */
+				<RA_PSEL(RA_PSEL_CEU, 4, 5)>, /* VIO_D2 */
+				<RA_PSEL(RA_PSEL_CEU, 4, 6)>, /* VIO_D3 */
+				<RA_PSEL(RA_PSEL_CEU, 7, 0)>, /* VIO_D4 */
+				<RA_PSEL(RA_PSEL_CEU, 7, 1)>, /* VIO_D5 */
+				<RA_PSEL(RA_PSEL_CEU, 7, 2)>, /* VIO_D6 */
+				<RA_PSEL(RA_PSEL_CEU, 7, 3)>, /* VIO_D7 */
+				<RA_PSEL(RA_PSEL_CEU, 7, 8)>, /* VIO_CLK */
+				<RA_PSEL(RA_PSEL_CEU, 7, 9)>, /* VIO_HD */
+				<RA_PSEL(RA_PSEL_CEU, 7, 10)>; /* VIO_VD */
+			drive-strength = "high";
+		};
+	};
```
Also, since the connector is already defined in the board dts, should we move this to shield layer?

khoa-nguyen-18 replied:
If this pinctrl definition is placed in the board layer, I think it would be nice to include `/omit-if-no-ref/`

thenguyenyf replied:
It not necessary in this case. `/omit-if-no-ref/` just have meaning in case you have multiple configuration for the pinctrl and actually use one in application. But with EK-RA8D1, the camera shield should only use the dvp_20pin connector on board.

khoa-nguyen-18 commented on dts/arm/renesas/ra/ra8/r7fa8d1xh.dtsi:
```diff
@@ -295,6 +295,13 @@
+		ceu: ceu@40348000 {
+			compatible = "renesas,ra-ceu";
+			reg = <0x40348000 0x8000>;
+			clocks = <&pclka MSTPC 16>;
```
I think it's missing `clock-names` here

khoa-nguyen-18 commented on boards/renesas/ek_ra8d1/doc/index.rst:
```diff
@@ -119,6 +119,14 @@ Supported Features
 
+   - For using the Camera Expansion Port (J59) with the Camera, please set switch SW1 as following configuration:
+
+     +-------------+-------------+--------------+------------+------------+------------+-------------+-----------+
+     | SW1-1 PMOD1 | SW1-2 TRACE | SW1-3 CAMERA | SW1-4 ETHA | SW1-5 ETHB | SW1-6 GLCD | SW1-7 SDRAM | SW1-8 I3C |
+     +-------------+-------------+--------------+------------+------------+------------+-------------+-----------+
+     |     OFF     |     OFF     |      ON     |     OFF    |     OFF    |     OFF     |     ON      |    OFF    |
+     +-------------+-------------+--------------+------------+------------+------------+-------------+-----------+
```
Should we just set the OFF state for the conflicting SW and "don't care" state for the non-conflicting SW?

<img width="982" height="440" alt="image" src="https://github.com/user-attachments/assets/1f15bb72-a38e-432c-b13c-2ed8669ff302" />

khoa-nguyen-18 reviewed (commented):
I have some comments regarding camera support on the board and in the shield layer

thenguyenyf reviewed (dismissed):
Just a comment for nit. Others LGTM.

josuah commented:
To summarize what allows this PR to be merged in practice:

- The "change requests" are both to add this board to the "build-all" target so that CI keeps compiling the driver making sure it works in the future [(comment)](https://github.com/zephyrproject-rtos/zephyr/pull/92146#pullrequestreview-3010258739)

- A git conflict preventing it to be merged.

Then of course letting time for PR author to check all comments.

leiche-cid commented:
Hi, thanks for your work, any plans to move this forward?

quytranpzz committed:
tests: drivers: build_all: Add Renesas RA video driver build

Enable video driver build for Renesas RA8D1

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

josuah commented on drivers/video/video_renesas_ra_ceu.c:
```diff
@@ -0,0 +1,540 @@
+	struct video_buffer *next_vbuf;
+	fsp_err_t err;
+
+	if (p_args->event & CEU_EVENT_FRAME_END) {
+		curr_vbuf = atomic_ptr_get(&data->vbuf);
+		if (curr_vbuf && atomic_ptr_cas(&data->vbuf, curr_vbuf, NULL)) {
+			curr_vbuf->timestamp = k_uptime_get_32();
+			k_fifo_put(&data->fifo_out, curr_vbuf);
+		}
```
If wanting a complete support for the polling API, you may want to raise a completion signal here, or remove support for the polling API altogether (given this will be automated soon).

https://github.com/zephyrproject-rtos/zephyr/blob/5525941d839099309a44a0be868151fa894a5425/drivers/video/video_sw_generator.c#L312-L314

josuah commented on drivers/video/video_renesas_ra_ceu.c:
```diff
@@ -0,0 +1,540 @@
+		err = R_CEU_CaptureStart(data->fsp_ctrl, next_vbuf->buffer);
+		if (err != FSP_SUCCESS) {
+			atomic_ptr_clear(&data->vbuf);
+			k_fifo_put(&data->fifo_out, next_vbuf);
```
If wanting a complete support for the polling API, you may want to raise a error signal here, or remove support for the polling API altogether (given this will be automated soon).

https://github.com/zephyrproject-rtos/zephyr/blob/5525941d839099309a44a0be868151fa894a5425/drivers/video/video_mcux_csi.c#L55

https://github.com/zephyrproject-rtos/zephyr/blob/5525941d839099309a44a0be868151fa894a5425/drivers/video/video_esp32_dvp.c#L116

josuah reviewed (approved):
I do not find anything blocking anymore.

Waiting on feedback from other reviewers who might want some modifications done before or after merging...

I would still recommend to:

- Remove the POLL API, which is optional, and currently incompletely supported. This intermediate situation would lead to some video samples to work (capture) and others to not work (UVC webcam).
- Open an issue for filtering the video formats of the source to only support the formats allowed by this device, as soon as the API making it is introduced (no RFC for it yet).
- Consider moving the call to enable the clock in the `_init()` function, if this is possible and makes sense in your opiinion.

Thank you for the modifications and for your patience!

thenguyenyf commented:
Hello @ngphibang . All change requests have been addressed. Still remain small nits as your comment but it can be sent out as a subsequence PR. Could you please revisit and verify it?

ngphibang reviewed (approved):
Thank you for this contribution !

There are a few reviewer comments that haven’t been addressed yet, though none are blocking. For future PRs, it might be helpful to share more of your thoughts on the commented points — it can really help reviewers better understand your perspective.
